File: COMPSYS 301/psoc student pack/psoc student pack/matlab/astar.m
%% A*'s Algorithm
% see astar explaination.txt for further detail

function [retmap,retvisited,retsteps] = astar(mapfile,startlocation,targetlocation)

    %list of squares we checked to find the nearest path    open list (queue)
    %list that we have decided and will not change again close list(visitedlist & retvisted)
    queue  = {startlocation}; % list of squares we checked to find the nearest path (open list)
    mapSize = size(mapfile);
    retmap = mapfile;
    retsteps = {};
    visitedList = {}; % list that we have decided and will not change again (closed list)
    retvisited = ones(mapSize); % map of visited nodes where 1 = not visited, 0 = visited
    parents = {}; % map of parent locations (ie: the neighbouring nodes have "parent" nodes --> parent nodes is the current node)
    
    % maps of H, G and F scores
    Hscore = zeros(mapSize);
    Fscore = zeros(mapSize);
    Gscore = zeros(mapSize);
    
    prevFscore = 2000;
    currentGcost = 0;
    stepNum = 0;
    visiting = [-1,-1]; % just for initalisation purposes
    checkifAtEnd = 0;


    while (checkifAtEnd ~= 1) % ends when we found target location
        
        % Find node with lowest Fscore to visit
        for i = 1:length(queue)
            location = queue{1,i};
            Frow = location(1);
            Fcols = location(2);
            nodeFscore = Fscore(Frow,Fcols);

            % check if node has the lowest Fscore
            if nodeFscore < prevFscore
                smallestFNode = location;
                index = i;
                prevFscore = nodeFscore;
            end
        end
            
        % visit smallest Fscore node and remove from queue
        visiting = smallestFNode;
        stepNum = stepNum + 1;
        prevFscore = 5000; % reset to checking previous F node value for next iteration
        visitedList= append(visitedList,visiting); %add to closed list
        retvisited(visiting(1),visiting(2)) = 0; % place visited on visited map
        queue = removeElement(queue,index);

        %checks if location of node is at end
        checkifAtEnd =(targetlocation(1) == visiting(1)) && (targetlocation(2) == visiting(2));
        
       
        % determine neighbours of visiting node and caluclate Fscore
         ycurrent = visiting(1);
         xcurrent = visiting(2); 

         % Neighbours are S, N,E, W from the current location
         neighbours = {[ycurrent+1,xcurrent],[ycurrent-1,xcurrent ],[ycurrent,xcurrent+1],[ycurrent,xcurrent-1]};
         
         % calculate Fscore if neighbour is not in closed list (visitedList --> ie visted array)
         % and not in open list (queue) and is not a wall
        for j = 1: length(neighbours)
            adjacentpos = neighbours{1,j};
            rows = adjacentpos(1);
            cols = adjacentpos(2);
            newFscore = 0;
            currentGcost = Gscore(ycurrent,xcurrent);
            if (rows > 0 && cols >0 && retmap(rows,cols) ~= 1 && retvisited(rows,cols) ~= 0) % not a wall and has not been visited
                
                % check if open list (queue) does not contain current neighbouring position
                if(contains(queue,adjacentpos) == false)
                    queue = append(queue,adjacentpos);
                    
                    % calculate Fscore, Gscore and Hscore
                    % Fscore = Gscore + Hscore
                    Gscore = calculateGScore(Gscore,adjacentpos,currentGcost);
                    Hscore = calculateHScore(Hscore,adjacentpos, targetlocation);
                    Fscore(rows,cols) = Gscore(rows,cols) + Hscore(rows,cols);
                    
                    parents{rows,cols} = [ycurrent,xcurrent]; % add parent for the neighbouring node
                
                else % if it is already in open list update fscore if needed and update parent
                    oldFscore = Fscore(rows,cols);
                    Gscore = calculateGScore(Gscore,adjacentpos,currentGcost);
                    Hscore = calculateHScore(Hscore,adjacentpos, targetlocation);
                    newFscore = Gscore(rows,cols) + Hscore(rows,cols);
                  
                    % check if new Fscore is smaller than old Fscore
                    if(newFscore < oldFscore)
                        Fscore(rows,cols) = newFscore;
                        parents{rows,cols} = [ycurrent,xcurrent]; % update parent node of neighbour to current node
                    end
                    
                end
            end
        end
    end

    % Finds the actual shortest path by going in the reverse direction and looking at
    % a neighbour's parent location. eg: target location is [1,18], the
    % parent of that point is [2,18] so we go to [2,18] and find it's parents
    % until we get to the start node.
    % Starts from target node and get the paths back to start node
    
    currentNode = targetlocation;
    parentNode = [-1,-1]; % initialised parent node
    reversedPath = {};
    checkStart = 0;
    reversedPath = append(reversedPath,currentNode); % add target to reversed path list
    
    while(checkStart ~= 1)
        parentNode = parents{currentNode(1),currentNode(2)}; % find parent node associated with current node
        reversedPath = append(reversedPath,parentNode); % add to reversed path list
        currentNode = parentNode; % for next iteration
        checkStart = (currentNode(1) == startlocation(1) && currentNode(2)== startlocation(2));
    end
    retsteps = flip(reversedPath); % flips order of elements to make a path to go from start to target. 
    

    % display path on map
    for k = 1:length(retsteps)
        placestep(retsteps{k},k);
    end
end


%% calculating H score
% H score is the estimated cost of movement from start point to target and
% ignores walls.
% //Previously used Manhattan Distance [Hmaps(xloc,yloc)=abs(xloc-target(1)) + abs(yloc-target(2));)]//
% Using Euclidean Distance (supposed to be used for when we can move in any
% direction but it seems to give a more accurate cost which in turn will
% generate a more accurate shortest path) We have tried use Manhattan Distance
% since its best used for only moving left, right,up and down but it did
% not give an accurate shortest path in some situations.
% Eclidian Distance --> H = square root ((currentx - targetx)^2 + (currenty-targety)^2)
% See: https://www.geeksforgeeks.org/a-search-algorithm/]
function [Hmaps] = calculateHScore(Hmap,current, target)
    xloc = current(1);
    yloc = current(2);
    Hmaps = Hmap;
    Hmaps(xloc,yloc) = sqrt((xloc-target(1))^2 + (yloc-target(2))^2);
    
    
end

%% calculating G score
% G score is the movement score from the start point to current point. We
% add to the current cost by one each time we move away from start point.
function [Gmaps] = calculateGScore(Gmap,currentpos,currentcost)
        Gmaps = Gmap;
        xloc = currentpos(1);
        yloc = currentpos(2);
        Gmaps(xloc,yloc) = currentcost +1;
end

%% Checks if array contains a location
function [doesContain] = contains(cellarray, target)
    doesContain = false;
    for i = 1:length(cellarray)
        pos = cellarray{1,i};
        if (pos == target)
            doesContain = true;
            break
        end
    end
end

%% removes an specified element in the array
function[array]= removeElement(poparray,index)
    array = poparray;
    % Input index has to be a integer
        array(index)= [];
    

end
%% Appends to end of array
function [updatedStack] = append(stack,node)
    updatedStack = stack;
    updatedStack{length(stack)+1} = node;

end

%% Placing steps on map
function placestep(position,i)
    % This function will plot a insert yellow rectangle and also print a number in this rectangle. Use with plotmap/viewmap. 
    position = [16-position(1) position(2)];
    position=[position(2)+0.1 position(1)+0.1];
    rectangle('Position',[position,0.8,0.8],'FaceColor','y');
    c=sprintf('%d',i);
    text(position(1)+0.2,position(2)+0.2,c,'FontSize',10);
end

File: COMPSYS 301/psoc student pack/psoc student pack/psoc_code_base/301-Astar-code/astar.c
#include "astar.h"

// --- Priority queue implementation modified from https://www.programiz.com/dsa/priority-queue
void swap(Pairs*a, Pairs *b) {
  Pairs temp = *b;
  *b = *a;
  *a = temp;
}

// Heapify the queue in order of minimum at the top of queue
void heapify(Paths array, int size, int i) {
  if (size == 1) {
    printf("Single element in the heap");
  } else {
    // Find the smallest value
    int smallest = i;
    int l = 2 * i + 1;
    int r = 2 * i + 2;

    // Use locations at smallest, L and R to find smallest Fscore in nodesArray
    int l_row = array[l].pair.pairRow;
    int l_col = array[l].pair.pairCol;
    int small_row = array[smallest].pair.pairRow;
    int small_col = array[smallest].pair.pairCol;
    int r_row = array[r].pair.pairRow;
    int r_col = array[r].pair.pairCol;
    
    if(l_row < MAP_ROWS && l_col < MAP_COLS && r_row < MAP_ROWS && r_col < MAP_COLS && l_row > 0 && l_col >0 && r_col>0 && r_row>0){

      // Find Fscores
      int l_fscore = nodesArray[l_row][l_col].fScore;
      int r_fscore = nodesArray[r_row][r_col].fScore;
      int small_fscore = nodesArray[small_row][small_col].fScore;

      // Compare Fscores to find smallest
      if (l < size && l_fscore < small_fscore)
        smallest = l;
      if (r < size && r_fscore < small_fscore)
        smallest = r;


      // Swap and continue heapifying if root is not smallest
      if (smallest != i) {
        swap(&array[i].pair, &array[smallest].pair);
        heapify(array, size, smallest);
      }
    }
  }
}

// Function to insert an element into the tree
void enqueue(Paths array, Pairs newLoc) {
  if (size == 0) {
    array[0].pair = newLoc;
    size += 1;
  } else if(size < 0){
    array[0].pair = newLoc;
    size = 0;
  }
  else {
    array[size].pair = newLoc;
    size += 1;
    for (int i = size / 2 - 1; i >= 0; i--) {
      heapify(array, size, i);
    }
  }
}

// Function to remove first element of the queues
Pairs dequeue(Paths array) {
    Pairs locAtTop;
    locAtTop.pairRow = array[0].pair.pairRow;
    locAtTop.pairCol = array[0].pair.pairCol;
    printf("Top of queue: %d, %d \n", locAtTop.pairRow, locAtTop.pairCol);

    swap(&array[0].pair, &array[size - 1].pair);
    size -= 1;
    if (size < 0){
      size = 0;
    }
    for (int i = size / 2 - 1; i >= 0; i--) {
      heapify(array, size, i);
    }

    return locAtTop;
}
//--- End of priority queue ---
void intialiseNodes(int mazeMap[MAP_ROWS][MAP_COLS]){
    for(int row=0; row<MAP_ROWS; row++){
        for(int col=0; col<MAP_COLS; col++){
            nodesArray[row][col].parents.pairRow = -1;
            nodesArray[row][col].parents.pairCol = -1;

            if(mazeMap[row][col] == 1){ // Wall
                nodesArray[row][col].traversable = 0;
            }else{
                nodesArray[row][col].traversable = 1; // traversable path
            }
            nodesArray[row][col].inVisited =  0;
            nodesArray[row][col].inQueue =  0;
            nodesArray[row][col].fScore = 2000;
            nodesArray[row][col].gScore = 0;
            nodesArray[row][col].hScore = 0;
        }
    }
}

// calculating H score using Manhattan Distance 
void calculateHScore(Pairs current, Pairs target){
    int xloc = current.pairCol;
    int yloc = current.pairRow;
    nodesArray[yloc][xloc].hScore = abs(xloc-target.pairRow) + abs(yloc-target.pairCol);
}

// Calculating G score  
// G score is the movement score from the start point to current point. We
// add to the current cost by one each time we move away from start point.
void calculateGScore(Pairs currentpos,int currentcost){
    int xloc = currentpos.pairCol;
    int yloc = currentpos.pairRow;
    nodesArray[yloc][xloc].gScore = currentcost +1;
}

void astar(int mapmaze[MAP_ROWS][MAP_COLS], int startLocRow,int startLocCol, int endLocRow, int endLocCol){
    Paths queue;
    Paths visitedList;
    Paths steps;

    intialiseNodes(mapmaze);
    Pairs target, start;
    target.pairRow = endLocRow-1;
    target.pairCol = endLocCol-1;
    start.pairCol = startLocRow-1;
    start.pairRow = startLocRow-1;
    
    queue[0].pair.pairRow = startLocRow-1;
    queue[0].pair.pairCol = startLocCol-1;
    int currentGcost = 0;
    int stepNum = 0;
    Pairs visiting, northNeigh,eastNeigh,southNeigh,westNeigh;
    visiting.pairRow = -1;
    visiting.pairCol = -1;
    int currentVisitedIndex = 0;
    int currentRow, currentCol;

   while(!((visiting.pairRow == target.pairRow) && (visiting.pairCol == target.pairCol))){
      printf("Size of queue: %d \n", size);
      visiting =  dequeue(queue);
      printf("Size of queue after dequeue: %d \n", size);
      printf("visiting (in matlab format [col][row]): %d,%d\n", visiting.pairCol+1,visiting.pairRow+1);
      stepNum += 1;
      // Append to visitied list
      visitedList[currentVisitedIndex].pair.pairRow = visiting.pairRow;
      visitedList[currentVisitedIndex].pair.pairCol = visiting.pairCol;
      nodesArray[visiting.pairRow][visiting.pairCol].inVisited = 1; // set flag for being in visited list
      currentVisitedIndex++;
      
      // determine neighbours of visiting node and caluclate Fscore
      currentCol = visiting.pairCol;
      currentRow = visiting.pairRow;

      northNeigh.pairRow = currentRow-1;
      northNeigh.pairCol = currentCol;
      southNeigh.pairRow = currentRow+1;
      southNeigh.pairCol= currentCol;
      eastNeigh.pairRow = currentRow;
      eastNeigh.pairCol = currentCol+1;
      westNeigh.pairRow = currentRow;
      westNeigh.pairCol = currentCol-1;
      
      // Neighbours are S, N,E, W from the current location
      Paths neighbours = {southNeigh,northNeigh,eastNeigh,westNeigh};

      //calculate Fscore if neighbour is not in closed list (visitedList --> ie visted array)
      // and not in open list (queue) and is not a wall
      for(int i=0; i<4;i++){
        Pairs adjacentPos; 
        adjacentPos = neighbours[i].pair;
        int row = adjacentPos.pairRow;
        int col = adjacentPos.pairCol;
        int newFscore = 0;
        int oldFscore = 2000;
        int currentGScore = nodesArray[row][col].gScore;
        //printf("Adjacent: %d,%d\n", adjacentPos.pairCol,adjacentPos.pairRow);
        
        //not a wall and has not been visited
        if((row > 0)&&(col>0) && (row < MAP_ROWS) && (col<MAP_COLS)){
          if((nodesArray[row][col].traversable == 1) && (nodesArray[row][col].inVisited == 0)){
            // check if open list (queue) does not contain current neighbouring position
            if(nodesArray[row][col].inQueue == 0){
              printf("Enqueuing: %d,%d\n",col,row);
              //calculate Fscore, Gscore and Hscore
              // Fscore = Gscore + Hscore
              calculateHScore(adjacentPos,target);
              calculateGScore(adjacentPos,currentGcost);
              nodesArray[row][col].fScore = nodesArray[row][col].gScore + nodesArray[row][col].hScore;

              enqueue(queue,adjacentPos);
              nodesArray[row][col].inQueue == 1;
              
              nodesArray[row][col].parents.pairRow = visiting.pairRow; // add parent for the neighbouring node
              nodesArray[row][col].parents.pairCol = visiting.pairCol;

            }else{// if it is already in open list update fscore if needed and update parent
              oldFscore = nodesArray[row][col].fScore;
              calculateHScore(adjacentPos,target);
              calculateGScore(adjacentPos,currentGcost);
              newFscore = nodesArray[row][col].gScore + nodesArray[row][col].hScore;

              //check if new Fscore is smaller than old Fscore, iif so update Fscore value
              if(newFscore < oldFscore){
                  nodesArray[row][col].fScore = newFscore;
                  nodesArray[row][col].parents.pairCol = visiting.pairCol; // update parent node of neighbour to current node
                  nodesArray[row][col].parents.pairRow = visiting.pairRow;
              }
            }
            printf("Adjacent: (%d,%d), Parents: (%d,%d)\n", adjacentPos.pairCol+1,adjacentPos.pairRow+1,nodesArray[row][col].parents.pairCol+1,nodesArray[row][col].parents.pairRow+1);
          }
        }
      }
   }

    
    struct Pair currentNode;
    struct Pair parentNode;
    Paths reversedArray;
    int index = 0;
    currentNode.pairCol = target.pairCol;
    currentNode.pairRow = target.pairRow;
    reversedArray[index].pair = target;
    index = 1;

    while(!(currentNode.pairRow == start.pairRow && currentNode.pairCol == start.pairCol)){
      //printf("current reversed Nodes: (%d,%d)\n", currentNode.pairCol+1,currentNode.pairRow+1);
      parentNode = nodesArray[currentNode.pairRow][currentNode.pairCol].parents;
      //printf("parent Node: (%d,%d)\n",parentNode.pairCol+1,parentNode.pairRow+1);

      reversedArray[index].pair = parentNode;
      currentNode.pairCol = parentNode.pairCol;
      currentNode.pairRow = parentNode.pairRow;
      index++;

    }

    
    int i = 0;
    printf("Final Path\n");
    for(int j=index; j--; j>=0){
      Pairs node = reversedArray[j].pair;
      finalPath[i].pair = node;
      printf("(%d,%d)\n", node.pairCol+1,node.pairRow+1);
      i++;
    }



      
    // currentNode = targetlocation;
    // parentNode = [-1,-1]; % initialised parent node
    // reversedPath = {};
    // checkStart = 0;
    // reversedPath = append(reversedPath,currentNode); % add target to reversed path list
    
    // while(checkStart ~= 1)
    //     parentNode = parents{currentNode(1),currentNode(2)}; % find parent node associated with current node
    //     reversedPath = append(reversedPath,parentNode); % add to reversed path list
    //     currentNode = parentNode; % for next iteration
    //     checkStart = (currentNode(1) == startlocation(1) && currentNode(2)== startlocation(2));
    // end
    // retsteps = flip(reversedPath); % flips order of elements to make a path to go from start to target. 
    
    
}
